numpy/core/function_base.py

```diff
@@ -356,3 +356,153 @@ def geomspace(start, stop, num=50, endpoint=True, dtype=None):
+        If true and `refpoint` is set to None, start, stop, or a special
+        combination of parameters then `stop` is the last sample.  Regardless,
+        if `endpoint` is set to True then the last sample will be the closest
+        number to `stop` on the floating-point scale within the specified
+        interval.  Otherwise, it will be the next closest number.
+        Default is True.
+    retscale : boolean, optional
+        If True, return (`samples`, `scale`), where `scale` is the multiplier
+        to convert from the unbiased floating-point values to the specified
+        interval.
+        Default is False.
+    dtype : dtype
+        The type of the output array.  If `dtype` is not given, infer the data
+        type from the other input arguments.
+        Default is None.
+
+    Returns
+    -------
+    samples : ndarray
+        samples spaced on a floating-point scale centered about `refpoint`
+    scale : float, optional
+        Only returned if `retscale` is True.  The multiplier to convert from
+        the unbiased floating-point values to the specified interval.
+
+    Examples
+    --------
+    >>> np.fpspace(0.0, 1.0, exponentbits=2, significandbits=0)
+    array([ 0., 0.25, 0.375, 0.4375, 0.5, 0.5625, 0.625, 0.75, 1. ])
+    >>> np.fpspace(0.0, 1.0, exponentbits=1, significandbits=1)
+    array([ 0., 0.125, 0.25, 0.375, 0.5, 0.625, 0.75, 0.875, 1. ])
+    >>> np.fpspace(0.0, 1.0, refpoint=0.0, exponentbits=2, significandbits=0)
+    array([ 0, 0.125, 0.25, 0.5, 1. ])
+    >>> np.fpspace(0.0, 1.0, refpoint=0.0, exponentbits=1, significandbits=1)
+    array([ 0., 0.25, 0.5, 0.75, 1. ])
+    >>> np.fpspace(0.0, 1.0, refpoint=0.0, exponentbits=1, significandbits=1, subnormals=False)
+    array([ 0., 0.5, 0.75, 1. ])
+
+    Graphical illustration:
+
+    >>> import matplotlib.pyplot as plt
+    >>> significandbits = 4
+    >>> x1 = np.fpspace(0.1, 1, significandbits=significandbits, subnormals=True)
+    >>> x2 = np.fpspace(0.1, 1, significandbits=significandbits, subnormals=False)
+    >>> y1 = np.zeros(x1.size)
+    >>> y2 = np.zeros(x2.size)
+    >>> plt.plot(x1, y1, 'o')
+    [<matplotlib.lines.Line2D object at 0x...>]
+    >>> plt.plot(x2, y2 + 0.5, 'o')
+    [<matplotlib.lines.Line2D object at 0x...>]
+    >>> plt.ylim([-0.5, 1])
+    (-0.5, 1)
+    >>> plt.show()
+
+    """
+    # Convert float/complex array scalars to float, gh-3504
+    # and make sure one can use variables that have an __array_interface__, gh-6634
+    start = asanyarray(start) * 1.0
+    stop  = asanyarray(stop)  * 1.0
+    exponentbits = int(exponentbits)
+    significandbits = int(significandbits)
+
+    dt = result_type(start, stop)
+    if dtype is None:
+        dtype = dt
+
+    if refpoint is None:
+        refpoint = 0.5 * (start + stop)
+
+    u_start = 0
+    u_stop = _nx.power(base, significandbits) - 1
+    u_num = u_stop - u_start + 1
+    u = linspace(u_start, u_stop, num=u_num, endpoint=True, dtype=dt)
```

Comment:
If you look at your test failures in Travis-CI, it's happening on this line, because `u_num` is a float, not an integer:
```
======================================================================
ERROR: numpy.core.tests.test_function_base.TestFpspace.test_dtype
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/numpy/numpy/builds/venv/lib/python3.4/site-packages/numpy/core/function_base.py", line 15, in _index_deprecate
    i = operator.index(i)
TypeError: 'numpy.float64' object cannot be interpreted as an integer
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/home/travis/build/numpy/numpy/builds/venv/lib/python3.4/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/travis/build/numpy/numpy/builds/venv/lib/python3.4/site-packages/numpy/core/tests/test_function_base.py", line 348, in test_dtype
    y = fpspace(0, 6, dtype='float32')
  File "/home/travis/build/numpy/numpy/builds/venv/lib/python3.4/site-packages/numpy/core/function_base.py", line 476, in fpspace
    u = linspace(u_start, u_stop, num=u_num, endpoint=True, dtype=dt)
  File "/home/travis/build/numpy/numpy/builds/venv/lib/python3.4/site-packages/numpy/core/function_base.py", line 101, in linspace
    num = _index_deprecate(num)
  File "/home/travis/build/numpy/numpy/builds/venv/lib/python3.4/site-packages/numpy/core/function_base.py", line 21, in _index_deprecate
    warnings.warn(msg, DeprecationWarning, stacklevel=stacklevel)
DeprecationWarning: object of type <class 'numpy.float64'> cannot be safely interpreted as an integer.
```

NumPy's test suite turns warnings issued internally into errors to ensure that we don't use deprecated behavior.

Reply:
Thank you for pointing that out!  It behaves like gcc's -Werror and so an explicit cast to integer should do the trick.